src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -22,6 +23,10 @@ internal sealed class AdditionalSourcesCollection
 
         private static readonly StringComparer s_hintNameComparer = StringComparer.OrdinalIgnoreCase;
 
+        // Matches relative path segments ("../", "./", "//"), "/" at the beginning,
+        // and " /" (directories ending with space cause problems).
+        private static readonly Regex s_invalidSegmentPattern = new Regex(@"(\.{1,2}|/|^| )/", RegexOptions.Compiled);
```

Comment:
curious why we're doing regex here, instead of using FS apis (that we already have) or things like looking for the current Path-Separator.  It always worries me if we're doing this sort of detection/munging ourselves.

Reply:
I haven't found FS API to do what I want - detect relative path segments (like `../`) in a given path. Note that this regex is only used to do such validation. For all path manipulations, normal FS APIs are used.